### PR TITLE
chapter hash

### PR DESCRIPTION
**Pull request by chachaleo**
Chapter about Hash 

Resolves : #397

### CODE REVIEW

**Review by enitrat — COMMENTED** [review]
great work :)

File: src/ch11-03-hash.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+The core library makes it easy to work with hashes. The `Hash` trait is implemented for all types that can be converted to `felt252`, including `felt252` itself. For more complex types like structs, deriving `Hash` allows them to be hashed easily using the hash function of your choice - given that all of the struct's fields are themselves hashable. You cannot derive the `Hash` trait on a struct that contains un-hashable values, such as `Array<T>` or a `Felt252Dict<T>`.
+
+The `Hash` trait is accompanied by the `HashStateTrait`, defined as follows:
+
+```rust
+{{#include ../listings/ch11-advanced-features/no_listing_03_hash_trait/src/lib.cairo:hashtrait}}
+```
+
+
+To use hashes in your code, you must first import the relevant traits and functions. In the following example, we will demonstrate how to hash a struct using both the Pedersen and Poseidon hash functions.
```
[the remark this answers is not in the document]
Comment (chachaleo) [reply src/ch11-03-hash.md:31]: `poseidon_hash_span` is not part of any Trait and is only implemented in the poseidon file. So I needed to import it. But maybe I missed something.

File: src/ch11-03-hash.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+#### When to use them ?
+
+Pedersen was introduced first, and is still used to compute the addresses of variables in storage (for example, `LegacyMap` uses Pedersen to hash the keys of a storage mapping on Starknet). As Poseidon is cheaper and faster than Pedersen when working with a STARK proof system, it should be used in priority when possible.
+
+### Working with Hashes
+
+The core library makes it easy to work with hashes. The `Hash` trait is implemented for all types that can be converted to `felt252`, including `felt252` itself. For more complex types like structs, deriving `Hash` allows them to be hashed easily using the hash function of your choice - given that all of the struct's fields are themselves hashable. You cannot derive the `Hash` trait on a struct that contains un-hashable values, such as `Array<T>` or a `Felt252Dict<T>`.
```
[the remark this answers is not in the document]
Comment (chachaleo) [reply src/ch11-03-hash.md:22]: Yes, I'll add the precision

File: src/ch11-03-hash.md [hunk 3]
```diff
@@ -19,9 +19,9 @@ Pedersen was introduced first, and is still used to compute the addresses of var
...
+The core library makes it easy to work with hashes. The `Hash` trait is implemented for all types that can be converted to `felt252`, including `felt252` itself. For more complex types like structs, deriving `Hash` allows them to be hashed easily using the hash function of your choice - given that all of the struct's fields are themselves hashable. You cannot derive the `Hash` trait on a struct that contains un-hashable values, such as `Array<T>` or a `Felt252Dict<T>`, even if `T`is hashable.
 
-The `Hash` trait is accompanied by the `HashStateTrait`, defined as follows:
+The `Hash` trait is accompanied by the `HashStateTrait` that defines the basic methods to work with hashes. They allow you to initialize a hash state, that contains the temporary values of the varibles involved in the computation of the hash. `HashStateTrait` also defines the methods to update the hash state and finalized it when the computation is completed. `HashStateTrait` is defined as follows:
```
Comment (enitrat) [inline src/ch11-03-hash.md:24]:
```suggestion
The `Hash` trait is accompanied by the `HashStateTrait` that defines the basic methods to work with hashes. They allow you to initialize a hash state that will contain the temporary values of the hash after each application of the hash function; update the hash state, and finalize it when the computation is completed. `HashStateTrait` is defined as follows:
```

**Review by enitrat — APPROVED** [review]
LGTM

**Pull request merged**